### ISSUE TITLE
Why is JSON RPC not used?

### ISSUE THREAD

**Issue by felixfbecker** [Microsoft/vscode-debugadapter-node#58]
Just wondering why the language server protocol uses JSON RPC, but the debug protocol uses structures like these
```ts
/** Base class of requests, responses, and events. */
    interface ProtocolMessage {
        /** Sequence number */
        seq: number;
        /** One of "request", "response", or "event" */
        type: string;
    }
    /** Client-initiated request */
    interface Request extends ProtocolMessage {
        /** The command to execute */
        command: string;
        /** Object containing arguments for the command */
        arguments?: any;
    }
```
that is basically JSON RPC, just with different property names (`seq -> id`, `arguments -> params`, `command -> method`). It would make implementations of debug adapters in other languages much easier because many languages have some kind of JSON RPC implementation.

**Comment by daviwil** [comment]
Agreed on this, I've had to do a fair amount of work to support both the language and debug server protocols with one class design in my C# code.  Would be nice if they could both share the same core protocol shape.

**Comment by felixfbecker** [comment]
I think there definitely is the need for an easy-to-implement, standardized, open source debug protocol. There are efforts in the wild but most of them depend on XML which makes them very verbose. This would of course be a breaking change for implementations, but it would only be a couple of properties renamed. Future implementations would have it much much easier in turn. Afaik the protocol is not used yet by a lot of IDEs and it is better to do this change now than later. For most VS Code extensions, the change would not be noticeable if they use vscode-debugadapter-node. JSON RPC has a `jsonrpc` field that identifies a message as a JSON RPC message, and the language server protocol uses a `Content-Type` header, so VS Code could even support both in a transition phase. You could also share a lot of code between the language server/client and the debug adapter/client, as they implemented the jsonrpc implementation in a separate package..

Proprietary protocols is not what people expect from the "new" Microsoft 😉

**Comment by weinand** [comment]
The base portion of the VS Code Debug Protocol is identical to the [V8 Debugging Protocol](https://github.com/v8/v8/wiki/Debugging-Protocol). When we started using it, it was not deprecated and many client libraries existed (or still exist). Since we did not invent it, I would not call it a 'proprietary MS protocol'.

Since it was a lean protocol designed for debugging, we considered this a good choice at that time. JSON RPC existed, but it was not prevalent in that area. Since quite a few debug adapters were using it when we shipped VS Code 1.0, we decided not not change the protocol so close to the first public release.

The initial version of the language server protocol was a variant of V8 Debugging Protocol. Because the language server work started a year later than the debugger, only very few clients existed and we decided to break them by switching to JSON RPC just before the VS Code 1.0 release. The major reason for changing the protocol was the fact that using a "debugging" protocol for a language server sounded strange.

We are considering switching to JSON RPC for the 2.0 version of the protocol (because that version will be breaking anyway).
